### DISCUSSION TITLE
Unable to verify the first certificate

### DISCUSSION THREAD

**Discussion by ssharma2089** [renovatebot/renovate#12355] [Request Help] [answered]
I am integrating renovatebot with bit-bucket server. However, I am getting unable to verify certificate and unknown bitbucket init Repo.
Below are the logs

DEBUG: Parsing configs
DEBUG: File config
       "config": {
         "platform": "bitbucket-server",
         "username": "sagar",
         "password": "***********",
         "onboarding": true,
         "gitAuthor": "Renovate Bot <bot@renovateapp.com>",
         "onboardingConfig": {"extends": ["config:base"]},
         "repositories": ["MAD/notification-common"],
         "hostRules": [{"matchHost": "git.xxx.xxx", "authType": "Basic"}]
       }
DEBUG: CLI config
       "config": {}
DEBUG: Env config
       "config": {
         "hostRules": [],
         "baseDir": "/tmp/renovate/",
         "platform": "bitbucket-server",
         "endpoint": "https://git.xxx.xxx/",
         "autodiscover": false
       }
DEBUG: Combined config
       "config": {
         "platform": "bitbucket-server",
         "username": "sagar",
         "password": "***********",
         "onboarding": true,
         "gitAuthor": "Renovate Bot <bot@renovateapp.com>",
         "onboardingConfig": {"extends": ["config:base"]},
         "repositories": ["MAD/notification-common"],
         "hostRules": [{"matchHost": "git.xxx.xxx", "authType": "Basic"}],
         "baseDir": "/tmp/renovate/",
         "endpoint": "https://git.xxx.xxx/",
         "autodiscover": false
       }
DEBUG: Adding trailing slash to endpoint
DEBUG: Using configured gitAuthor (Renovate Bot <bot@renovateapp.com>)
DEBUG: Adding password authentication for git.xxx.xxx to hostRules
DEBUG: Using configured baseDir: /tmp/renovate/
DEBUG: Using cacheDir: /tmp/renovate/cache
DEBUG: Initializing Renovate internal cache into /tmp/renovate/cache/renovate/renovate-cache-v1
DEBUG: Commits limit = null
DEBUG: Clearing hostRules
DEBUG: Adding password authentication for git.xxx.xxx to hostRules
 INFO: Repository started (repository=MAD/notification-common)
       "renovateVersion": "28.10.1"
DEBUG: Using localDir: /tmp/renovate/repos/bitbucket-server/MAD/notification-common (repository=MAD/notification-common)
DEBUG: initRepo("{
         "repository": "MAD/notification-common"
       }") (repository=MAD/notification-common)
DEBUG: Unknown Bitbucket initRepo error (repository=MAD/notification-common)
       "err": {
         "name": "RequestError",
         "code": "UNABLE_TO_VERIFY_LEAF_SIGNATURE",
         "timings": {
           "start": 1635333139032,
           "socket": 1635333139036,
           "lookup": 1635333139038,
           "connect": 1635333139040,
           "error": 1635333139053,
           "phases": {"wait": 4, "dns": 2, "tcp": 2, "total": 21}
         },
         "message": "unable to verify the first certificate",
         "stack": "RequestError: unable to verify the first certificate\n    at ClientRequest.<anonymous> (/usr/src/app/node_modules/got/dist/source/core/index.js:956:111)\n    at Object.onceWrapper (events.js:520:26)\n    at ClientRequest.emit (events.js:412:35)\n    at ClientRequest.emit (domain.js:475:12)\n    at ClientRequest.origin.emit (/usr/src/app/node_modules/@szmarczak/http-timer/dist/source/index.js:43:20)\n    at TLSSocket.socketErrorListener (_http_client.js:475:9)\n    at TLSSocket.emit (events.js:400:28)\n    at TLSSocket.emit (domain.js:475:12)\n    at emitErrorNT (internal/streams/destroy.js:106:8)\n    at emitErrorCloseNT (internal/streams/destroy.js:74:3)\n    at processTicksAndRejections (internal/process/task_queues.js:82:21)\n    at TLSSocket.onConnectSecure (_tls_wrap.js:1515:34)\n    at TLSSocket.emit (events.js:400:28)\n    at TLSSocket.emit (domain.js:475:12)\n    at TLSSocket._finishInit (_tls_wrap.js:937:8)\n    at TLSWrap.ssl.onhandshakedone (_tls_wrap.js:709:12)",
         "options": {
           "headers": {
             "user-agent": "RenovateBot/28.10.1 (https://github.com/renovatebot/renovate)",
             "x-atlassian-token": "no-check",
             "authorization": "***********",
             "accept": "application/json",
             "accept-encoding": "gzip, deflate, br"
           },
           "url": "https://git.xxx.xxx/rest/api/1.0/projects/MAD/repos/notification-common",
           "hostType": "bitbucket-server",
           "username": "",
           "password": "",
           "method": "GET",
           "http2": false
         }
       }
ERROR: Repository has unknown error (repository=MAD/notification-common)
       "err": {
         "name": "RequestError",
         "code": "UNABLE_TO_VERIFY_LEAF_SIGNATURE",
         "timings": {
           "start": 1635333139032,
           "socket": 1635333139036,
           "lookup": 1635333139038,
           "connect": 1635333139040,
           "error": 1635333139053,
           "phases": {"wait": 4, "dns": 2, "tcp": 2, "total": 21}
         },
         "message": "unable to verify the first certificate",
         "stack": "RequestError: unable to verify the first certificate\n    at ClientRequest.<anonymous> (/usr/src/app/node_modules/got/dist/source/core/index.js:956:111)\n    at Object.onceWrapper (events.js:520:26)\n    at ClientRequest.emit (events.js:412:35)\n    at ClientRequest.emit (domain.js:475:12)\n    at ClientRequest.origin.emit (/usr/src/app/node_modules/@szmarczak/http-timer/dist/source/index.js:43:20)\n    at TLSSocket.socketErrorListener (_http_client.js:475:9)\n    at TLSSocket.emit (events.js:400:28)\n    at TLSSocket.emit (domain.js:475:12)\n    at emitErrorNT (internal/streams/destroy.js:106:8)\n    at emitErrorCloseNT (internal/streams/destroy.js:74:3)\n    at processTicksAndRejections (internal/process/task_queues.js:82:21)\n    at TLSSocket.onConnectSecure (_tls_wrap.js:1515:34)\n    at TLSSocket.emit (events.js:400:28)\n    at TLSSocket.emit (domain.js:475:12)\n    at TLSSocket._finishInit (_tls_wrap.js:937:8)\n    at TLSWrap.ssl.onhandshakedone (_tls_wrap.js:709:12)",
         "options": {
           "headers": {
             "user-agent": "RenovateBot/28.10.1 (https://github.com/renovatebot/renovate)",
             "x-atlassian-token": "no-check",
             "authorization": "***********",
             "accept": "application/json",
             "accept-encoding": "gzip, deflate, br"
           },
           "url": "https://git.xxx.xxx/rest/api/1.0/projects/MAD/repos/notification-common",
           "hostType": "bitbucket-server",
           "username": "",
           "password": "",
           "method": "GET",
           "http2": false
         }
       }
DEBUG: Unknown res (repository=MAD/notification-common)
       "res": "unknown-error"
DEBUG: Repository timing splits (milliseconds) (repository=MAD/notification-common)
       "splits": {},
       "total": 75
 INFO: Repository finished (repository=MAD/notification-common)
       "durationMs": 75
DEBUG: Renovate exiting

**Comment by rarkins** [comment, marked as answer]
Looks like your bitbucket uses a self-signed certificate, which Node.js doesn't trust.

**Reply by viceice** [reply]
Adding custom ca cert to renovate: https://github.com/containerbase/buildpack/blob/main/docs/custom-root-ca.md#runtime-install

**Reply by ssharma2089** [reply]
@rarkins @viceice I am deploying renovate as cronjob(kubernetes). Instead of adding crt in buildtime. Is there any env variable to add in cronjob to read crt. 

Referring below link for kubernetes
https://github.com/renovatebot/renovate/blob/main/docs/usage/examples/self-hosting.md

**Reply by rarkins** [reply]
I don't think so. How do you do that with other containers?

**Reply by viceice** [reply]
@ssharma2089 check my link, there is a sample to add cert via emv on runtime

**Reply by ssharma2089** [reply]
@rarkins @viceice I am getting below error "Cannot connect to maven host". Could you please help me to resolve this

DEBUG: Content is not found for Maven url: https://repo.maven.apache.org/maven2/io/micronaut/micronaut-parent/maven-metadata.xml (repository=xxx/notification-xxx)
DEBUG: Failed to look up dependency io.micronaut:micronaut-parent (repository=xxx/notification-xxx, packageFile=pom.xml, dependency=io.micronaut:micronaut-parent)
DEBUG: Looking up commons-io:commons-io in repository https://repo.maven.apache.org/maven2/ (repository=xxx/notification-xxx)
DEBUG: Cannot connect to maven host (repository=xxx/xxx)
       "failedUrl": "https://repo.maven.apache.org/maven2/org/apache/logging/log4j/log4j-api/maven-metadata.xml"

**Reply by rarkins** [reply]
It can't connect to the maven server

**Reply by ssharma2089** [reply]
@rarkins How to connect to maven server. Is there any env variable which I need configure to connect.

**Reply by rarkins** [reply]
It normally just works, so you must have some network blocking or proxy causing the problem

**Reply by ssharma2089** [reply]
@rarkins I am running behind corporate firewall. Is there way to define HTTPS host to env variable. So that it can connect to that proxy

**Reply by viceice** [reply]
You can use normal linux proxy variables, renovate should detect those

**Reply by ssharma2089** [reply]
@viceice If I add below env variable host and port then it will work or not
- image: renovate/renovate:28.10.1
              name: renovate-bot
              env: # For illustration purposes, please use secrets.
                - name: RENOVATE_PLATFORM
                  value: 'bitbucket-server'
                - name: RENOVATE_AUTODISCOVER
                  value: 'false'
	       - name: HTTP_PROXY_HOST
                  value: "xxx.xxx.int"
                - name: HTTP_PROXY_PORT
                  value: "3128"

**Reply by viceice** [reply]
nope, use those standard vars:
https://github.com/renovatebot/renovate/blob/dd462359b41b80ebbdd40d22a6ac63d0cb0e9acc/lib/proxy.ts#L4-L4

**Reply by ssharma2089** [reply]
ok. i will try that

**Reply by ssharma2089** [reply]
@viceice After adding above env variable. getting below error

DEBUG: Unknown Bitbucket initRepo error (repository=xxx/notification-xxx)
       "err": {
         "name": "RequestError",
         "code": "EPROTO",
         "timings": {
           "start": 1636110968280,
           "socket": 1636110968292,
           "lookup": 1636110968292,
           "connect": 1636110968292,
           "upload": 1636110968294,
           "error": 1636110968296,
           "phases": {"wait": 12, "dns": 0, "tcp": 0, "request": 2, "total": 16}
         },
         "message": "write EPROTO 140362550675328:error:1408F10B:SSL routines:ssl3_get_record:wrong version number:../deps/openssl/openssl/ssl/record/ssl3_record.c:332:\n",
         "stack": "RequestError: write EPROTO 140362550675328:error:1408F10B:SSL routines:ssl3_get_record:wrong version number:../deps/openssl/openssl/ssl/record/ssl3_record.c:332:\n\n    at ClientRequest.<anonymous> (/usr/src/app/node_modules/got/dist/source/core/index.js:956:111)\n    at Object.onceWrapper (events.js:520:26)\n    at ClientRequest.emit (events.js:412:35)\n    at ClientRequest.emit (domain.js:475:12)\n    at ClientRequest.origin.emit (/usr/src/app/node_modules/@szmarczak/http-timer/dist/source/index.js:43:20)\n    at TLSSocket.socketErrorListener (_http_client.js:475:9)\n    at TLSSocket.emit (events.js:412:35)\n    at TLSSocket.emit (domain.js:475:12)\n    at emitErrorNT (internal/streams/destroy.js:106:8)\n    at emitErrorCloseNT (internal/streams/destroy.js:74:3)\n    at processTicksAndRejections (internal/process/task_queues.js:82:21)\n    at WriteWrap.onWriteComplete [as oncomplete] (internal/stream_base_commons.js:94:16)",

**Reply by viceice** [reply]
It seems your proxy has some invalid ssl config, as this error is thrown by openssl lib, which is used by nodejs under the hood.

https://github.com/openssl/openssl/issues/6289

You should try to do some curl tests with your proxy to see, if requests passthrough.
